presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java

```diff
@@ -115,32 +124,32 @@ else if (expression instanceof DereferenceExpression) {
 
     public Optional<ResolvedField> tryResolveField(Expression node, QualifiedName name)
     {
-        return resolveField(node, name, true);
+        return resolveField(node, name, 0, true);
     }
 
-    private Optional<ResolvedField> resolveField(Expression node, QualifiedName name, boolean local)
+    private Optional<ResolvedField> resolveField(Expression node, QualifiedName name, int fieldIndexOffset, boolean local)
     {
         List<Field> matches = relation.resolveFields(name);
         if (matches.size() > 1) {
             throw ambiguousAttributeException(node, name);
         }
         else if (matches.size() == 1) {
-            return Optional.of(asResolvedField(getOnlyElement(matches), local));
+            return Optional.of(asResolvedField(getOnlyElement(matches), fieldIndexOffset, local));
         }
         else {
             if (isColumnReference(name, relation)) {
                 return Optional.empty();
             }
             if (parent.isPresent()) {
-                return parent.get().resolveField(node, name, false);
+                return parent.get().resolveField(node, name, fieldIndexOffset + relation.getAllFieldCount(), local && !queryBoundary);
             }
             return Optional.empty();
         }
     }
 
-    private ResolvedField asResolvedField(Field field, boolean local)
+    private ResolvedField asResolvedField(Field field, int fieldIndexOffset, boolean local)
     {
-        int fieldIndex = relation.indexOf(field);
+        int fieldIndex = relation.indexOf(field) + fieldIndexOffset;
```

Comment:
I don't doubt you needed a change here. I think, however, that such semantic change could trigger field rename to better accommodate the new meaning. And to avoid creating a trap.